### DISCUSSION TITLE
Zwift + FTMS Server

### DISCUSSION THREAD

**Discussion by MarkusSchneider** [doudar/SmartSpin2k#296] [Show and tell]
Sequence of FTMS-Server commands send from Zwift in ERG Mode:
1) "Control Request" (0x00)
2) "Reset" (0x01)
3) "Control Request" (0x00)
4) "Start Training" (0x07)
5 ...) "Set Target Power" (0x05)
6) If you pause Training -> "Set Indoor Bike Simulation Parameter" (0x11) why is Zwift not sending "Stop or Pause" (0x08) ??
7) If you resume Training -> "Set Target Power" (0x05)
8) If workout is finished -> "Set Indoor Bike Simulation Parameter" (0x11) 

Sequence of FTMS-Server commands send from Zwift in Sim Mode:
1) "Control Request" (0x00)
2) "Reset" (0x01)
3) "Control Request" (0x00)
4) "Start Training" (0x07)
5..) "Set Indoor Bike Simulation Parameter" (0x11) 

Even if Game is ended Zwift does not send "Stop or Pause"

**Comment by doudar** [comment]
It is a bit weird but I’m guessing it’s because Zwift is otherwise always in Sim mode. 

You’re right, we should implement stop though.

**Reply by MarkusSchneider** [reply]
Create a new Branch and added Stop and Cadence Control Point.
Is WIP, not sure if everything is korrekt.
See https://github.com/doudar/SmartSpin2k/blob/2d3b3db62d9839453c47b9f7c3dce756c77806d0/src/BLE_Server.cpp#L72-L79
and https://github.com/doudar/SmartSpin2k/blob/2d3b3db62d9839453c47b9f7c3dce756c77806d0/src/BLE_Server.cpp#L422-L433
and https://github.com/doudar/SmartSpin2k/blob/2d3b3db62d9839453c47b9f7c3dce756c77806d0/src/BLE_Server.cpp#L467-L481

Don't know if line 475 is correct. 
For what is the fitnessMachineStatusCharacteristic used? I can not find a description of this value in the FTMS specification.

**Comment by doudar** [comment]
It's on 4.17 (page66) and you'll see an example in Appendix 3 on the FTMS document for the spin down procedure. 

Then there's "training status" on page 46. 

It's a very confusing document

**Reply by doudar** [reply]
And it looks like training status could use some (re)work for each mode (case) as well.

I The first byte gets initialized to 0x08 on line 85 (which I don't think is correct) and never gets changed. 0x08 means "fitness test" if it was in byte 1 but nothing in byte 0. 

 Edit: I see some of those case statements are correct. We just need to change the initialization to 0x00, 0x00

**Comment by doudar** [comment]
I just noticed another small thing that should be fixed.

https://github.com/doudar/SmartSpin2k/blob/ceefb349b6f914751ff2aa509d2bc6c06047ea33/src/BLE_Server.cpp#L83

Shouldn't be hardcoded and should instead reference:

https://github.com/doudar/SmartSpin2k/blob/ceefb349b6f914751ff2aa509d2bc6c06047ea33/include/settings.h#L82

and

https://github.com/doudar/SmartSpin2k/blob/ceefb349b6f914751ff2aa509d2bc6c06047ea33/include/SmartSpin_parameters.h#L130

**Comment by MarkusSchneider** [comment]
According FitnessMachineTrainingStatus. Since SS2k is not controlling the training it's maybe not necessary to notify training status.
See FTMS Chapter 4.10.1 
![image](https://user-images.githubusercontent.com/7011066/147933172-cba33b6d-0869-41e4-8e28-da7d485d2bf0.png)

**Reply by doudar** [reply]
Maybe you should use it for the btsimulator.html 😝

**Comment by MarkusSchneider** [comment]
uint8_t cadenceStatus[3] = {0x15, (uint8_t)rxValue[2], (uint8_t)rxValue[1]}; 
fitnessMachineStatusCharacteristic->setValue(ERGStatus, 3); 

Should be correct, needs to be fixed for power value as well.

See FTMS Chapter 4.17Fitness Machine Status
![image](https://user-images.githubusercontent.com/7011066/147935001-c3a6f52c-7c2d-4f16-b31a-34d82dbce02f.png)

![image](https://user-images.githubusercontent.com/7011066/147935033-c78dd7e1-1122-446e-a6d6-bb72bcf7abfe.png)

OpCode -> 0x15: Target Cadence Changed
Parameter -> Target Cadence as read from Client
